kuzzleio/kuzzle
# KZL-1005 -  Add method to use request context in SDK queries

EDIT: following the discussion we had here about this functionality, I have to edit this PR but I keep the original one bellow.

## What does this PR do ?

Adds a method to use the original request user inside the requests sent by the embedded SDK.  
The main purpose of this feature is to properly set Kuzzle metadata (`author` and `updater`) on documents.

Call to `sdk.as(user)` method will instantiate a new SDK with a `FunnelProcotol` that takes the user in the constructor and inject it in the request context before passing the request to the`Funnel`.  

```
// This will create a document with `author` metadata to `null`
this.context.accessors.sdk.document.create(...); 

// This will create a document with `author` metadata to the original request user
this.context.accessors.sdk.as(request.context.user).document.create(...);

// This will throw a PluginImplementationError, no user impersonation for now
this.context.accessors.sdk.as({ user: { _id: 'gordon' } }).document.create(...);
```

Documentation: https://github.com/kuzzleio/documentation/pull/253

### How should this be manually tested?

Consider the following plugin called `ctoi-le-plugin`
```
class Plugin {
  constructor () {
    this.context = null;
  }

  init (config, context) {
    this.context = context;

    this.controllers = {
      test: {
        createDocs: 'createDocs'
      }
    };

    this.routes = [
      { verb: 'get', url: '/test/createDocs', controller: 'test', action: 'createDocs' }
    ];

    this.pipes = { }

    this.hooks = { }
  }

  async createDocs(request) {
    this.context.accessors.sdk.as(request.context.user).document.create('nyc-open-data', 'yellow-taxi', { toto: 'lol'});

    this.context.accessors.sdk.document.create('nyc-open-data', 'yellow-taxi', { toto: 'lol'});
    return true;
  }
}
```
  - Step 1 : Call the plugin route as anonymous, 2 document should be created, one will have `_kuzzle_info.author === '-1'` and the second `_kuzzle_info.author === null`


  - Step 1 : Call the plugin route [http://localhost:7512/_plugin/ctoi-le-plugin/test/createDocs](http://localhost:7512/_plugin/ctoi-le-plugin/test/createDocs)
  - Step 3 :  Call the plugin route, 2 document should be created, one will have `_kuzzle_info.author === 'admin'` and the second `_kuzzle_info.author === null`

## Other changes

 - throw an error when accessing the `sdk.realtime` controller

# ORIGINAL PR
## What does this PR do ?

When making a request with the embedded SDK, the executed request does not have the same context as the original request.  
As a result, the documents created do not have the `author` metadata for example.

This PR adds a `sdk.context()` method that take the original request JWT and send it along with the following SDK request. Queries that does not have a `sdk.context()` call before their execution will have the `author` metadata` set to `null`.  

Consider this controller action:
```
  async createDocs (request) {
    this.context.accessors.sdk.context(request).document.create('nyc-open-data', 'yellow-taxi', { foo: 'bar'})

    this.context.accessors.sdk.document.create('nyc-open-data', 'yellow-taxi', { foo: 'bar'})
  }
```
If I am logged as the admin user as I execute this controller action, the first call to `document.create` will result to a document with `author` metadata set to `'admin'`.  
The second call is not contextualized, so the document will be created with the `author` metadata set to `null`

### How should this be manually tested?

Consider the following plugin
```
class Plugin {
  constructor () {
    this.context = null;
  }

  init (config, context) {
    this.context = context;

    this.controllers = {
      test: {
        createDocs: 'createDocs'
      }
    };

    this.routes = [
      { verb: 'get', url: '/test/createDocs', controller: 'test', action: 'createDocs' }
    ];

    this.pipes = { }

    this.hooks = { }
  }

  async createDocs(request) {
    this.context.accessors.sdk.context(request).document.create('nyc-open-data', 'yellow-taxi', { toto: 'lol'});

    this.context.accessors.sdk.document.create('nyc-open-data', 'yellow-taxi', { toto: 'lol'});
    return true;
  }
}
```
  - Step 1 : Call the plugin route as anonymous, 2 document should be created, one will have `_kuzzle_info.author === '-1'` and the second `_kuzzle_info.author === null`


  - Step 1 : Log as admin
  - Step 3 :  Call the plugin route, 2 document should be created, one will have `_kuzzle_info.author === 'admin'` and the second `_kuzzle_info.author === null`
